### PR TITLE
Sampling profiler

### PR DESCRIPTION
**Pull request by abadams**
This is a sampling profiler for Halide. 

The current instrumenting profiler is annoying to use on Android because you can't read the cycle counter, so it injects a bunch of extra timing calls and really slows down your code (e.g. profiling HelloAndroidCamera2 is not really feasible). It also has no visibility into parallel loops.

Regular sampling profilers are great because they can be very accurate while minimally interfering with runtime. They're a pain because they're very OS-specific - they need to ask the OS what instruction pointer another process is at, and then figure out what that corresponds to in the source using debugging metadata.

This sampling profiler instruments the generated code to include a single extra store instruction (i.e. a movl) at every ProducerConsumer node. It stores an id of the currently running Func. It also launches a background profiler thread that periodically reads this id and uses it to keep statistics. It can for example accurately profile performance_fft, which uses tons of tiny stages. On process exit (or earlier - see HalideRuntime.h), it prints out stats. You turn it on with a target flag. Here's the output for local laplacian:
```
$ HL_TARGET=host-profile make out.png
make -C ../../ bin/libHalide.a
make[1]: Entering directory `/usr/local/google/home/abadams/projects/Halide'
make[1]: `bin/libHalide.a' is up to date.
make[1]: Leaving directory `/usr/local/google/home/abadams/projects/Halide'
g++-4.8 -std=c++11 -I../../include/ -I../support/ local_laplacian_gen.cpp -g ../../bin/libHalide.a -o local_laplacian_gen -lpthread -ldl -lz \
	 
./local_laplacian_gen
g++-4.8 -std=c++11 -I../../include/ -I../support/ -Wall -O3 process.cpp local_laplacian.o -o process -lpthread -ldl -L/usr/lib/x86_64-linux-gnu -lpng12 -I/usr/include/libpng12   
./process ../images/rgb.png 8 1 1 10 out.png
79796
68898
69512
69847
67203
71450
62430
70803
73233
70483
62430
local_laplacian  total time: 769.724304 ms  samples: 687  runs: 11  time per run: 69.974937 ms
  overhead:              0.000000ms     (0%)
  gray:                  9.673176ms     (13%)
  inGPyramid[1]:         4.378913ms     (6%)
  inGPyramid[2]:         0.973734ms     (1%)
  inGPyramid[3]:         0.586571ms     (0%)
  inGPyramid[4]:         0.489311ms     (0%)
  inGPyramid[5]:         0.000000ms     (0%)
  inGPyramid[6]:         0.000000ms     (0%)
  inGPyramid[7]:         0.000000ms     (0%)
  remap:                 0.000000ms     (0%)
  gPyramid[1]:           20.498981ms    (29%)
  gPyramid[2]:           10.608436ms    (15%)
  gPyramid[3]:           2.937738ms     (4%)
  gPyramid[4]:           0.682699ms     (0%)
  gPyramid[5]:           0.290462ms     (0%)
  gPyramid[6]:           0.291354ms     (0%)
  gPyramid[7]:           0.000000ms     (0%)
  outGPyramid[7]:        0.000000ms     (0%)
  outGPyramid[6]:        0.000000ms     (0%)
  outGPyramid[5]:        0.000000ms     (0%)
  outGPyramid[4]:        0.391489ms     (0%)
  outGPyramid[3]:        0.784669ms     (1%)
  outGPyramid[2]:        0.493306ms     (0%)
  outGPyramid[1]:        4.332160ms     (6%)
  outGPyramid[0]:        4.361609ms     (6%)
  local_laplacian:       8.200326ms     (11%)
```

This currently works on linux and android. Support for the other OSes is easy, but I thought I should see if people think this is even a good idea first. It also needs some dedicated tests (you can inspect the profiling state programmatically from the running program) and a tutorial that demonstrates how to use it.

Some of the code in the PR is a small refactor of how destructors work in order to distinguish between things that should always run no matter how a pipeline exits, and things that should only run if a pipeline exits on an error condition. This is relevant because you always want to store the id that says "I'm not inside a Halide pipeline right now" on pipeline exit whether it exits cleanly or not.

### CODE REVIEW

**Comment by steven-johnson** [comment]
> I thought I should see if people think this is even a good idea first

Anything that improves the profiling toolset for Halide is a big huge YES.

**Comment by zvookin** [comment]
I think this is a great idea. Only question is whether we're doing this in
addition to the existing instrumentation based profiler or instead of and
whether doing the oprofile support Dillon was looking at is still
worthwhile at some point. I don't think these decisions should prevent
merging this, but toward the goal of making them, I'd ask these questions:

1) What is the functionality restriction? E.g. the instrumented profiler
does not handle parallel. The oprofile based approach may have platforms it
does not work on.

2) What is the profiling overhead for each solution?

3) How accurate are the results?

I'll try to read the PR soon, but is the store into a pre-thread location?
Is there any advantage to keeping a small stack of the ids or using a
circular buffer of them. (There is no time info in the circular buffer so
it seems questionable, but it might give a bit more resolution.)

-Z-

**Comment by dsharletg** [comment]
I still want oprofile/other profiler support. Among other reasons:

- VTune at least provides a ton more information than just timing. oprofile can probably do this to. Memory access info, cache, etc. It can easily overwhelm you with data if you want it to.

- VTune/oprofile are probably limited in platform support. Though, last I looked, it at least looked plausibly supported on android.

That said, I think this PR is good and should be merged, pending code review of course.

**Comment by zvookin** [comment]
("pre-thread" should of course be "per-thread". I asked the question
because I think I understand how it handles parallel if the store is to a
"current Func for this thread" but I don't understand if it is global.)

**Comment by abadams** [comment]
It can handle parallel loops, but it can't profile inside of gpu loops,
because it's based on having a cheap store to a memory location that a CPU
thread can inspect. Maybe this can be made to work.

It does not use a separate location per thread, and instead just lets the
threads race each other to store their id. In halide each thread is doing
the same inner loop with the same sequence of Func realizations, so even if
one thread keeps winning the race there's no reason to think that this
would introduce any bias towards one Func over another. I thought a lot
about using small circular buffers, but it turns into a nasty constrained
optimization problem.

The accuracy is a function of how many samples you get. By default it's
configured to sample once every millisecond, so it's usefully accurate for
pipelines that you run for at least 100ms. Running it multiple times
increases accuracy proportionately. If you get a few thousand samples it's
extremely accurate. On my desktop this sampling frequency introduces no
visible overhead for compute_rooty pipelines like local_laplacian. The fft
is much more sensitive. It's single-threaded, so the cycles taken by the
profiler are meaningless, but storing to a cache line that another core
keeps reading from causes some sort of MESI protocol thrashing and you take
a 10% performance hit.

On android for the pipelines I've tested this sampling frequency seems to
cause a 10-20% performance hit too. You can lower the sampling frequency to
avoid this, at the expense of having to run your pipeline more times to get
enough samples.

This replaces the instrumenting profiler, and does not interfere with any
future oprofile solution.

**Comment by abadams** [comment]
This is ready to go. PTAL.

**Comment by abadams** [comment]
belay that, sync_fetch_and_add doesn't work on msvc. I should use a c++11 atomic.

**Comment by abadams** [comment]
Ok, figured out the (infuriating) windows issue. Should be good to go.

File: src/Lower.cpp [hunk 1]
```diff
@@ -58,7 +58,7 @@ using std::map;
 using std::pair;
 using std::make_pair;
 
-Stmt lower(const vector<Function> &outputs, const Target &t, const vector<IRMutator *> &custom_passes) {
+Stmt lower(const vector<Function> &outputs, const string &pipeline_name, const Target &t, const vector<IRMutator *> &custom_passes) {
```
Comment (dsharletg) [inline src/Lower.cpp]: Now that the pipeline name is here, I'd love to add it to the asserts that get generated. A consistent complaint I have is when I get a runtime assert failure without being able to tell which pipeline it came from. Maybe for another PR though.

**Comment by dsharletg** [comment]
LGTM, cool!

**Pull request merged**